File: benchmark/2026-09-17-7677581(latest)/llm-practical/trials/PE-P1-svm-specA/go/trial_01/solution.go
// WL-SVM / PE-P1-svm-specA — soft-margin SVM trained by projected dual ascent.
//
// Single file, standard library only (fmt, os, strings).
// No linear-algebra / tensor / statistics / BLAS facility is used: every
// accumulation below is a hand-written scalar loop (frozen rule C-9).
//
// Three logical modules, as required by spec 2.1:
//   data — the LCG-PM generator and the dataset builder
//   svm  — training, support-vector extraction, w/b recovery, f, g, evaluation
//   app  — entry point and output formatting
package main

import (
	"bufio"
	"fmt"
	"os"
)

// ---------------------------------------------------------------------------
// Frozen constants (spec 2.2)
// ---------------------------------------------------------------------------

const (
	D          = 4
	NTrainC1   = 100
	NTrainC2   = 100
	NTestC1    = 50
	NTestC2    = 50
	Sigma      = 0.8
	Seed       = 1234567
	C          = 10.0
	LR         = 0.0001
	Limit      = 0.0001
	Sweeps     = 1000
	EpsSV      = 0.0000001
	NTrain     = NTrainC1 + NTrainC2 // 200
	NTest      = NTestC1 + NTestC2   // 100
	SVMVersion = 1
)

var (
	MU1 = [D]float64{1.0, 1.0, 0.5, -0.5}
	MU2 = [D]float64{-1.0, -1.0, -0.5, 0.5}
)

// ===========================================================================
// module: data
// ===========================================================================

// lcgPM is the Park–Miller minimal-standard LCG (cross-language constraint C-1):
//
//	state_{n+1} = (state_n * 48271) mod 2147483647
//	unit_float  = state_{n+1} / 2147483647.0
//
// The largest intermediate, 48271 * 2147483646 ~ 1.037e14, is exact in int64.
type lcgPM struct {
	state int64
}

func newLCGPM(seed int64) *lcgPM {
	return &lcgPM{state: seed}
}

// nextUnit advances the stream and returns the next unit float in (0,1).
func (r *lcgPM) nextUnit() float64 {
	r.state = (r.state * 48271) % 2147483647
	return float64(r.state) / 2147483647.0
}

// nextNormal returns an Irwin–Hall(12) standard normal deviate:
// the sum of twelve consecutive unit floats, minus six.
func (r *lcgPM) nextNormal() float64 {
	s := 0.0
	for k := 0; k < 12; k++ {
		s = s + r.nextUnit()
	}
	return s - 6.0
}

// fillBlock draws n points of D dimensions around mu, points ascending and,
// within a point, dimensions ascending.
func fillBlock(r *lcgPM, n int, mu [D]float64) [][D]float64 {
	block := make([][D]float64, n)
	for i := 0; i < n; i++ {
		for d := 0; d < D; d++ {
			block[i][d] = mu[d] + Sigma*r.nextNormal()
		}
	}
	return block
}

type dataset struct {
	xTrain []([D]float64) // 200 training points, class +1 block then class -1 block
	yTrain []float64      // +1 for i < 100, -1 for i >= 100
	xTest1 []([D]float64) // 50 test points of class +1
	xTest2 []([D]float64) // 50 test points of class -1
}

// buildDataset draws the four blocks from one stream, in the exact order of
// spec 2.3, and concatenates the training blocks class +1 first.
func buildDataset() *dataset {
	r := newLCGPM(Seed)

	train1 := fillBlock(r, NTrainC1, MU1)
	train2 := fillBlock(r, NTrainC2, MU2)
	test1 := fillBlock(r, NTestC1, MU1)
	test2 := fillBlock(r, NTestC2, MU2)

	xTrain := make([][D]float64, 0, NTrain)
	yTrain := make([]float64, 0, NTrain)
	for i := 0; i < NTrainC1; i++ {
		xTrain = append(xTrain, train1[i])
		yTrain = append(yTrain, 1.0)
	}
	for i := 0; i < NTrainC2; i++ {
		xTrain = append(xTrain, train2[i])
		yTrain = append(yTrain, -1.0)
	}

	return &dataset{xTrain: xTrain, yTrain: yTrain, xTest1: test1, xTest2: test2}
}

// ===========================================================================
// module: svm
// ===========================================================================

type model struct {
	alpha        []float64
	beta         float64
	judge        bool
	errorLast    float64
	maxAbsDelta  float64
	w            [D]float64
	b            float64
	nsMargin     int
	nsInside     int
	sMargin      []int
	sInside      []int
	objective    float64
	alphaSum     float64
	alphaYSum    float64
	alphaChecksm float64
}

// gram builds the Gram matrix once. Precomputing is mandatory (spec 2.4).
func gram(x [][D]float64) [][]float64 {
	n := len(x)
	g := make([][]float64, n)
	for i := 0; i < n; i++ {
		g[i] = make([]float64, n)
	}
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			s := 0.0
			for d := 0; d < D; d++ {
				s = s + x[i][d]*x[j][d]
			}
			g[i][j] = s
		}
	}
	return g
}

// train runs exactly SWEEPS full sweeps of projected dual ascent, with no
// early exit. alpha is updated in place (Gauss–Seidel), beta once per sweep.
func train(g [][]float64, y []float64) *model {
	n := len(y)

	alpha := make([]float64, n)
	for i := 0; i < n; i++ {
		alpha[i] = 0.0
	}
	beta := 1.0

	judge := false
	errAcc := 0.0
	maxAbsDelta := 0.0

	for sweep := 0; sweep < Sweeps; sweep++ {

		judge = false
		errAcc = 0.0
		maxAbsDelta = 0.0

		// (3.1) update alpha, ascending i, in place
		for i := 0; i < n; i++ {

			item1 := 0.0
			for j := 0; j < n; j++ {
				item1 = item1 + alpha[j]*y[i]*y[j]*g[i][j]
			}

			item2 := 0.0
			for j := 0; j < n; j++ {
				item2 = item2 + alpha[j]*y[i]*y[j]
			}

			delta := 1.0 - item1 - beta*item2

			ad := delta
			if ad < 0.0 {
				ad = -ad
			}
			if ad > maxAbsDelta {
				maxAbsDelta = ad
			}

			alpha[i] = alpha[i] + LR*delta
			if alpha[i] < 0.0 {
				alpha[i] = 0.0
			} else if alpha[i] > C {
				alpha[i] = C
			} else if ad > Limit {
				judge = true
				errAcc = errAcc + (ad - Limit)
			}
		}

		// (3.2) update beta, once per sweep
		s := 0.0
		for i := 0; i < n; i++ {
			s = s + alpha[i]*y[i]
		}
		beta = beta + s*s/2.0
	}

	return &model{
		alpha:       alpha,
		beta:        beta,
		judge:       judge,
		errorLast:   errAcc,
		maxAbsDelta: maxAbsDelta,
	}
}

// recover extracts the support-vector index sets and rebuilds w and b.
func (m *model) recover(x [][D]float64, y []float64) {
	n := len(y)

	sMargin := make([]int, 0, n)
	sInside := make([]int, 0, n)
	for i := 0; i < n; i++ {
		if EpsSV < m.alpha[i] && m.alpha[i] < C-EpsSV {
			sMargin = append(sMargin, i)
		}
	}
	for i := 0; i < n; i++ {
		if m.alpha[i] >= C-EpsSV {
			sInside = append(sInside, i)
		}
	}
	m.sMargin = sMargin
	m.sInside = sInside
	m.nsMargin = len(sMargin)
	m.nsInside = len(sInside)

	for d := 0; d < D; d++ {
		m.w[d] = 0.0
	}
	for d := 0; d < D; d++ {
		for _, i := range sMargin {
			m.w[d] = m.w[d] + m.alpha[i]*y[i]*x[i][d]
		}
		for _, i := range sInside {
			m.w[d] = m.w[d] + m.alpha[i]*y[i]*x[i][d]
		}
	}

	b := 0.0
	for _, i := range sMargin {
		dp := 0.0
		for d := 0; d < D; d++ {
			dp = dp + m.w[d]*x[i][d]
		}
		b = b + (y[i] - dp)
	}
	if m.nsMargin == 0 {
		fmt.Fprintln(os.Stderr, "FAIL: |S_margin| == 0")
		os.Exit(1)
	}
	m.b = b / float64(m.nsMargin)
}

// f is the decision function.
func (m *model) f(p [D]float64) float64 {
	s := 0.0
	for d := 0; d < D; d++ {
		s = s + m.w[d]*p[d]
	}
	return s + m.b
}

// g is the label predictor.
func (m *model) g(p [D]float64) int {
	if m.f(p) >= 0.0 {
		return 1
	}
	return -1
}

// summarize computes the dual objective and the alpha summaries.
func (m *model) summarize(gm [][]float64, y []float64) {
	n := len(y)

	sum := 0.0
	for i := 0; i < n; i++ {
		sum = sum + m.alpha[i]
	}

	quad := 0.0
	for i := 0; i < n; i++ {
		inner := 0.0
		for j := 0; j < n; j++ {
			inner = inner + m.alpha[i]*m.alpha[j]*y[i]*y[j]*gm[i][j]
		}
		quad = quad + inner
	}

	m.objective = sum - 0.5*quad
	m.alphaSum = sum

	ays := 0.0
	for i := 0; i < n; i++ {
		ays = ays + m.alpha[i]*y[i]
	}
	m.alphaYSum = ays

	cks := 0.0
	for i := 0; i < n; i++ {
		cks = cks + m.alpha[i]*float64((i%97)+1)
	}
	m.alphaChecksm = cks
}

// accuracy counts how many points of a block the model labels as want.
func (m *model) accuracy(block [][D]float64, want int) int {
	correct := 0
	for i := 0; i < len(block); i++ {
		if m.g(block[i]) == want {
			correct++
		}
	}
	return correct
}

// ===========================================================================
// module: app
// ===========================================================================

func main() {
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	ds := buildDataset()
	gm := gram(ds.xTrain)
	m := train(gm, ds.yTrain)
	m.recover(ds.xTrain, ds.yTrain)
	m.summarize(gm, ds.yTrain)

	converged := 1
	if m.judge {
		converged = 0
	}

	// Training accuracy over the 200 training points, class +1 block then
	// class -1 block (which is exactly index order).
	trainCorrect := 0
	for i := 0; i < NTrain; i++ {
		want := 1
		if ds.yTrain[i] < 0.0 {
			want = -1
		}
		if m.g(ds.xTrain[i]) == want {
			trainCorrect++
		}
	}
	trainAcc := float64(trainCorrect) / float64(NTrain)

	correctC1 := m.accuracy(ds.xTest1, 1)
	correctC2 := m.accuracy(ds.xTest2, -1)
	testCorrect := correctC1 + correctC2
	testAcc := float64(testCorrect) / float64(NTest)
	testAccC1 := float64(correctC1) / float64(NTestC1)
	testAccC2 := float64(correctC2) / float64(NTestC2)

	fmt.Fprintf(out, "SVM_VERSION %d\n", SVMVersion)
	fmt.Fprintf(out, "SWEEPS %d\n", Sweeps)
	fmt.Fprintf(out, "CONVERGED %d\n", converged)
	fmt.Fprintf(out, "ERROR_LAST %.6f\n", m.errorLast)
	fmt.Fprintf(out, "MAX_ABS_DELTA %.6f\n", m.maxAbsDelta)
	fmt.Fprintf(out, "BETA %.6f\n", m.beta)
	fmt.Fprintf(out, "NS_MARGIN %d\n", m.nsMargin)
	fmt.Fprintf(out, "NS_INSIDE %d\n", m.nsInside)
	fmt.Fprintf(out, "W %.6f %.6f %.6f %.6f\n", m.w[0], m.w[1], m.w[2], m.w[3])
	fmt.Fprintf(out, "B %.6f\n", m.b)
	fmt.Fprintf(out, "OBJECTIVE %.6f\n", m.objective)
	fmt.Fprintf(out, "ALPHA_SUM %.6f\n", m.alphaSum)
	fmt.Fprintf(out, "ALPHA_Y_SUM %.12f\n", m.alphaYSum)
	fmt.Fprintf(out, "ALPHA_CHECKSUM %.6f\n", m.alphaChecksm)
	fmt.Fprintf(out, "TRAIN_ACC %.6f\n", trainAcc)
	fmt.Fprintf(out, "TEST_ACC %.6f\n", testAcc)
	fmt.Fprintf(out, "TEST_ACC_C1 %.6f\n", testAccC1)
	fmt.Fprintf(out, "TEST_ACC_C2 %.6f\n", testAccC2)
	fmt.Fprintf(out, "TEST_CORRECT %d %d %d\n", correctC1, correctC2, testCorrect)

	out.WriteString("PRED")
	for i := 0; i < NTestC1; i++ {
		fmt.Fprintf(out, " %d", m.g(ds.xTest1[i]))
	}
	for i := 0; i < NTestC2; i++ {
		fmt.Fprintf(out, " %d", m.g(ds.xTest2[i]))
	}
	out.WriteString("\n")

	out.WriteString("ALPHA")
	for i := 0; i < NTrain; i++ {
		fmt.Fprintf(out, " %.6f", m.alpha[i])
	}
	out.WriteString("\n")
}

File: benchmark/2026-09-17-7677581(latest)/llm-practical/trials/PE-P1-svm-specA/zig/trial_01/solution.zig
// WL-SVM : SoftMargin-SVM dual ascent, spec PE-P1-svm-specA
// Logical modules: data, svm, app  (see section markers below)
// Standard library only. Single threaded. No linear-algebra / BLAS / stats facility.

const std = @import("std");

// ---------------------------------------------------------------------------
// frozen constants (2.2)
// ---------------------------------------------------------------------------
const D: usize = 4;
const N_TRAIN_C1: usize = 100;
const N_TRAIN_C2: usize = 100;
const N_TEST_C1: usize = 50;
const N_TEST_C2: usize = 50;
const N: usize = N_TRAIN_C1 + N_TRAIN_C2; // 200
const N_TEST: usize = N_TEST_C1 + N_TEST_C2; // 100

const MU1 = [D]f64{ 1.0, 1.0, 0.5, -0.5 };
const MU2 = [D]f64{ -1.0, -1.0, -0.5, 0.5 };
const SIGMA: f64 = 0.8;
const SEED: i64 = 1234567;
const C_BOUND: f64 = 10.0;
const LR: f64 = 0.0001;
const LIMIT: f64 = 0.0001;
const SWEEPS: usize = 1000;
const EPS_SV: f64 = 0.0000001;

const MULT: i64 = 48271;
const MOD: i64 = 2147483647;

// ===========================================================================
// module: data   -- LCG-PM generator and dataset builder
// ===========================================================================

var rng_state: i64 = SEED;

fn nextState() i64 {
    rng_state = @rem(MULT * rng_state, MOD);
    return rng_state;
}

fn nextUniform() f64 {
    const s: f64 = @floatFromInt(nextState());
    return s / 2147483647.0;
}

fn nextNormal() f64 {
    // Irwin-Hall(12) - 6 ; ascending accumulation, not reassociated
    var t: f64 = 0.0;
    var k: usize = 0;
    while (k < 12) : (k += 1) {
        t = t + nextUniform();
    }
    return t - 6.0;
}

var x_train: [N][D]f64 = undefined;
var y_train: [N]f64 = undefined;
var x_test: [N_TEST][D]f64 = undefined;
var y_test: [N_TEST]f64 = undefined;

fn fillBlock(dst: [][D]f64, start: usize, count: usize, mu: [D]f64) void {
    var n: usize = 0;
    while (n < count) : (n += 1) {
        var d: usize = 0;
        while (d < D) : (d += 1) {
            dst[start + n][d] = mu[d] + SIGMA * nextNormal();
        }
    }
}

fn buildDataset() void {
    // exact draw order (2.3)
    fillBlock(x_train[0..], 0, N_TRAIN_C1, MU1);
    fillBlock(x_train[0..], N_TRAIN_C1, N_TRAIN_C2, MU2);
    fillBlock(x_test[0..], 0, N_TEST_C1, MU1);
    fillBlock(x_test[0..], N_TEST_C1, N_TEST_C2, MU2);

    var i: usize = 0;
    while (i < N) : (i += 1) {
        y_train[i] = if (i < N_TRAIN_C1) 1.0 else -1.0;
    }
    i = 0;
    while (i < N_TEST) : (i += 1) {
        y_test[i] = if (i < N_TEST_C1) 1.0 else -1.0;
    }
}

// ===========================================================================
// module: svm
// ===========================================================================

var gram: [N][N]f64 = undefined;
var alpha: [N]f64 = undefined;
var beta: f64 = 1.0;

var last_judge: bool = false;
var last_error: f64 = 0.0;
var last_max_abs_delta: f64 = 0.0;

fn buildGram() void {
    var i: usize = 0;
    while (i < N) : (i += 1) {
        var j: usize = 0;
        while (j < N) : (j += 1) {
            var s: f64 = 0.0;
            var d: usize = 0;
            while (d < D) : (d += 1) {
                s = s + x_train[i][d] * x_train[j][d];
            }
            gram[i][j] = s;
        }
    }
}

fn train() void {
    var i: usize = 0;
    while (i < N) : (i += 1) {
        alpha[i] = 0.0;
    }
    beta = 1.0;

    var sweep: usize = 0;
    while (sweep < SWEEPS) : (sweep += 1) {
        var judge: bool = false;
        var err_acc: f64 = 0.0;
        var max_abs_delta: f64 = 0.0;

        // (3.1) alpha update, ascending i, in place (Gauss-Seidel)
        i = 0;
        while (i < N) : (i += 1) {
            var item1: f64 = 0.0;
            var j: usize = 0;
            while (j < N) : (j += 1) {
                item1 = item1 + alpha[j] * y_train[i] * y_train[j] * gram[i][j];
            }

            var item2: f64 = 0.0;
            j = 0;
            while (j < N) : (j += 1) {
                item2 = item2 + alpha[j] * y_train[i] * y_train[j];
            }

            const delta: f64 = 1.0 - item1 - beta * item2;
            const ad: f64 = @abs(delta);
            if (ad > max_abs_delta) max_abs_delta = ad;

            alpha[i] = alpha[i] + LR * delta;
            if (alpha[i] < 0.0) {
                alpha[i] = 0.0;
            } else if (alpha[i] > C_BOUND) {
                alpha[i] = C_BOUND;
            } else if (ad > LIMIT) {
                judge = true;
                err_acc = err_acc + (ad - LIMIT);
            }
        }

        // (3.2) beta update, once per sweep
        var s: f64 = 0.0;
        i = 0;
        while (i < N) : (i += 1) {
            s = s + alpha[i] * y_train[i];
        }
        beta = beta + s * s / 2.0;

        last_judge = judge;
        last_error = err_acc;
        last_max_abs_delta = max_abs_delta;
    }
}

var s_margin: [N]usize = undefined;
var ns_margin: usize = 0;
var s_inside: [N]usize = undefined;
var ns_inside: usize = 0;
var w: [D]f64 = undefined;
var b_bias: f64 = 0.0;

fn recover() void {
    ns_margin = 0;
    ns_inside = 0;
    var i: usize = 0;
    while (i < N) : (i += 1) {
        if (EPS_SV < alpha[i] and alpha[i] < C_BOUND - EPS_SV) {
            s_margin[ns_margin] = i;
            ns_margin += 1;
        }
        if (alpha[i] >= C_BOUND - EPS_SV) {
            s_inside[ns_inside] = i;
            ns_inside += 1;
        }
    }

    var d: usize = 0;
    while (d < D) : (d += 1) {
        w[d] = 0.0;
    }
    d = 0;
    while (d < D) : (d += 1) {
        var k: usize = 0;
        while (k < ns_margin) : (k += 1) {
            const idx = s_margin[k];
            w[d] = w[d] + alpha[idx] * y_train[idx] * x_train[idx][d];
        }
        k = 0;
        while (k < ns_inside) : (k += 1) {
            const idx = s_inside[k];
            w[d] = w[d] + alpha[idx] * y_train[idx] * x_train[idx][d];
        }
    }

    var bb: f64 = 0.0;
    var k: usize = 0;
    while (k < ns_margin) : (k += 1) {
        const idx = s_margin[k];
        var dp: f64 = 0.0;
        d = 0;
        while (d < D) : (d += 1) {
            dp = dp + w[d] * x_train[idx][d];
        }
        bb = bb + (y_train[idx] - dp);
    }
    const cnt: f64 = @floatFromInt(ns_margin);
    b_bias = bb / cnt;
}

fn fval(p: *const [D]f64) f64 {
    var s: f64 = 0.0;
    var d: usize = 0;
    while (d < D) : (d += 1) {
        s = s + w[d] * p[d];
    }
    return s + b_bias;
}

fn gval(p: *const [D]f64) f64 {
    return if (fval(p) >= 0.0) 1.0 else -1.0;
}

// ===========================================================================
// module: app
// ===========================================================================

var outbuf: [262144]u8 = undefined;
var outlen: usize = 0;

fn emit(comptime fmt: []const u8, args: anytype) void {
    const s = std.fmt.bufPrint(outbuf[outlen..], fmt, args) catch return;
    outlen += s.len;
}

fn nz(v: f64) f64 {
    return if (v == 0.0) 0.0 else v;
}

pub fn main() void {
    buildDataset();
    buildGram();
    train();
    recover();

    // aggregate quantities
    var alpha_sum: f64 = 0.0;
    var i: usize = 0;
    while (i < N) : (i += 1) {
        alpha_sum = alpha_sum + alpha[i];
    }

    var alpha_y_sum: f64 = 0.0;
    i = 0;
    while (i < N) : (i += 1) {
        alpha_y_sum = alpha_y_sum + alpha[i] * y_train[i];
    }

    var quad: f64 = 0.0;
    i = 0;
    while (i < N) : (i += 1) {
        var inner: f64 = 0.0;
        var j: usize = 0;
        while (j < N) : (j += 1) {
            inner = inner + alpha[i] * alpha[j] * y_train[i] * y_train[j] * gram[i][j];
        }
        quad = quad + inner;
    }
    const objective: f64 = alpha_sum - 0.5 * quad;

    var checksum: f64 = 0.0;
    i = 0;
    while (i < N) : (i += 1) {
        const m: f64 = @floatFromInt((i % 97) + 1);
        checksum = checksum + alpha[i] * m;
    }

    var train_correct: usize = 0;
    i = 0;
    while (i < N) : (i += 1) {
        if (gval(&x_train[i]) == y_train[i]) train_correct += 1;
    }

    var pred: [N_TEST]i64 = undefined;
    var correct_c1: usize = 0;
    var correct_c2: usize = 0;
    i = 0;
    while (i < N_TEST) : (i += 1) {
        const g = gval(&x_test[i]);
        pred[i] = if (g >= 0.0) 1 else -1;
        if (g == y_test[i]) {
            if (i < N_TEST_C1) correct_c1 += 1 else correct_c2 += 1;
        }
    }
    const total_correct: usize = correct_c1 + correct_c2;

    const train_acc: f64 = @as(f64, @floatFromInt(train_correct)) / @as(f64, @floatFromInt(N));
    const test_acc: f64 = @as(f64, @floatFromInt(total_correct)) / @as(f64, @floatFromInt(N_TEST));
    const test_acc_c1: f64 = @as(f64, @floatFromInt(correct_c1)) / @as(f64, @floatFromInt(N_TEST_C1));
    const test_acc_c2: f64 = @as(f64, @floatFromInt(correct_c2)) / @as(f64, @floatFromInt(N_TEST_C2));

    const converged: i64 = if (last_judge) 0 else 1;

    emit("SVM_VERSION 1\n", .{});
    emit("SWEEPS {d}\n", .{SWEEPS});
    emit("CONVERGED {d}\n", .{converged});
    emit("ERROR_LAST {d:.6}\n", .{nz(last_error)});
    emit("MAX_ABS_DELTA {d:.6}\n", .{nz(last_max_abs_delta)});
    emit("BETA {d:.6}\n", .{nz(beta)});
    emit("NS_MARGIN {d}\n", .{ns_margin});
    emit("NS_INSIDE {d}\n", .{ns_inside});

    emit("W", .{});
    var d: usize = 0;
    while (d < D) : (d += 1) {
        emit(" {d:.6}", .{nz(w[d])});
    }
    emit("\n", .{});

    emit("B {d:.6}\n", .{nz(b_bias)});
    emit("OBJECTIVE {d:.6}\n", .{nz(objective)});
    emit("ALPHA_SUM {d:.6}\n", .{nz(alpha_sum)});
    emit("ALPHA_Y_SUM {d:.12}\n", .{nz(alpha_y_sum)});
    emit("ALPHA_CHECKSUM {d:.6}\n", .{nz(checksum)});
    emit("TRAIN_ACC {d:.6}\n", .{nz(train_acc)});
    emit("TEST_ACC {d:.6}\n", .{nz(test_acc)});
    emit("TEST_ACC_C1 {d:.6}\n", .{nz(test_acc_c1)});
    emit("TEST_ACC_C2 {d:.6}\n", .{nz(test_acc_c2)});
    emit("TEST_CORRECT {d} {d} {d}\n", .{ correct_c1, correct_c2, total_correct });

    emit("PRED", .{});
    i = 0;
    while (i < N_TEST) : (i += 1) {
        emit(" {d}", .{pred[i]});
    }
    emit("\n", .{});

    emit("ALPHA", .{});
    i = 0;
    while (i < N) : (i += 1) {
        emit(" {d:.6}", .{nz(alpha[i])});
    }
    emit("\n", .{});

    const io = std.Options.debug_io;
    std.Io.File.stdout().writeStreamingAll(io, outbuf[0..outlen]) catch {};
}
